src/lightning/data/streaming/dataset.py

```diff
@@ -457,3 +405,59 @@ def is_integer(value: str) -> bool:
         return True
     except Exception:
         return False
+
+
+def _associate_chunks_to_workers(
+    num_workers: int, worker_env: _WorkerEnv, chunks_replica: List[int], intervals_replica: List[Any]
+) -> Any:
+    workers_chunks = {}
+    workers_intervals = {}
+
+    for worker_idx in range(num_workers):
+        worker_chunks = []
+        worker_intervals = []
+        for i, (chunk_index, chunk_interval) in enumerate(zip(chunks_replica, intervals_replica)):
+            if i % worker_env.world_size != worker_idx:
+                continue
+
+            worker_chunks.append(chunk_index)
+            worker_intervals.append(chunk_interval)
+
+        workers_chunks[worker_idx] = worker_chunks
+        workers_intervals[worker_idx] = worker_intervals
+
+    return workers_chunks, workers_intervals
+
+
+def _replay_sampling(num_samples_yielded: int, batch_size: int, num_workers: int) -> Dict[int, int]:
+    """This function replays the sampling from the dataloader."""
```

Comment:
I don't see any sampling happening in this function. The while loop below can be eliminated completely by setting the values directly, no?

Essentially, 

`indexes[worker_idx] = X * batch_size`, and all you need to compute is X, how many times does worker_idx get to process a batch.

Reply:
@awaelchli Thanks, this function was not optimized and would have been expensive with trillion of samples seen, I pushed a tiny update.

This `indexes[worker_idx] = X * batch_size` won't work but this would:

```python
    divisible_num_batches_yielded = num_samples_yielded // (num_workers * batch_size)

    indexes = {}
    for worker_idx in range(num_workers):
        indexes[worker_idx] = divisible_num_batches_yielded * batch_size

    num_samples_yielded = num_samples_yielded - (num_workers * divisible_num_batches_yielded * batch_size)
````

And then, we can take care of the reminder.